### PR TITLE
Block the hoisting of TYP_STRUCT rvalues in loop hoisting

### PR DESCRIPTION
**Pull request by briansull**
PR #22255  introduced test regressions:

`Assertion failed 'tree->gtHasReg()' in 'Program:Test(ref):this' (IL size 80)`

### CODE REVIEW

File: src/jit/optcse.cpp [hunk 1]
```diff
@@ -2574,11 +2574,29 @@ bool Compiler::optIsCSEcandidate(GenTree* tree)
...
+        if (structHnd == NO_CLASS_HANDLE)
+        {
+            return false;
+        }
+
+        unsigned          size = info.compCompHnd->getClassSize(structHnd);
+        structPassingKind howToPassStruct;
+        var_types         structBaseType = getArgTypeForStruct(structHnd, &howToPassStruct, false, size);
+
+        // Only allow structs whose storage can be held in a register to be CSE candidates.
```
Comment (mikedn) [inline src/jit/optcse.cpp]: Presumably that's just because it's not very useful to CSE something that doesn't fit in a register. Or is there a correctness issue as well?

I suppose it would make sense to CSE a struct indir if the new temporary variable can then be promoted. But that's probably quite complicated to get right.

Comment (tannergooding) [inline src/jit/optcse.cpp]: Is "register" referring just to integer registers and to the lower element of vector registers (for float/double)?

I imagine we may want to support CSE (and other optimizations) for full vector registers eventually...

**Comment by sandreenko** [comment]
cc @dotnet/jit-contrib.

@briansull is it possible to create a CoreClr repro based on desktop failures for this issue?

File: src/jit/optcse.cpp (hunk 1 above)
Comment (briansull) [reply src/jit/optcse.cpp]: Yes, There currently is a correctness issue as well when we perform loop hoisting we place a l-value which is a GT_IND of a TYP_STRUCT into the loop pre-header.
This indirection doesn't get handled properly (or removed) and we end up asserting about a missing register during codegen.

Comment (briansull) [reply src/jit/optcse.cpp]: see bool Compiler::optHoistLoopExprsForTree()
where we currenmtly require:

```
        // Tree must be a suitable CSE candidate for us to be able to hoist it.
        treeIsHoistable &= optIsCSEcandidate(tree);
```

Comment (briansull) [reply src/jit/optcse.cpp]: The call to `getArgTypeForStruct`  should identify the register sized SIMD struct types as 
` (howToPassStruct ==  SPK_PrimitiveType)`

Comment (briansull) [reply src/jit/optcse.cpp]: Currently it doesn't appear that we support passing such vector types in registers  according to `getArgTypeForStruct`

Comment (tannergooding) [reply src/jit/optcse.cpp]: I believe @CarolEidt is working on this for the Vector ABI support.

**Comment by briansull** [comment]
I will investigate creating a CoreCLR repro test case

File: src/jit/optcse.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/optcse.cpp]: Ah, of course, loop hoisting will hoist the LHS of the struct assignment but that does not have a struct handle so there's no way for it to actually work.

**Comment by mikedn** [comment]
Here's a repro in case you don't have one already:
```C#
    struct Str
    {
        float a, b, c, d, e, f;
    }

    class Cls
    {
        public Str s;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Sink<T>(ref T t)
    {
    }

    static void Main() => new Program().Test(new Cls());

    [MethodImpl(MethodImplOptions.NoInlining)]
    void Test(Cls c)
    {
        Str l1 = default;
        Str l2 = default;
        Str l3 = default;

        for (int i = 0; i < 10; i++)
        {
            l1 = c.s;
            l2 = c.s;
            l3 = c.s;
        }

        Sink(ref l1);
        Sink(ref l2);
        Sink(ref l3);
    }
```
asserts:
```
Assert failure(PID 14128 [0x00003730], Thread: 16408 [0x4018]): Assertion failed 'tree->gtHasReg()' in 'Program:Test(ref):this' (IL size 80)

    File: d:\projects\coreclr\src\jit\codegenlinear.cpp Line: 1289
    Image: D:\Projects\coreclr\bin\Product\Windows_NT.x64.Checked\CoreRun.exe
```

**Comment by briansull** [comment]
Thank you mikedn

**Comment by briansull** [comment]
Test case added

**Comment by briansull** [comment]
PTAL @dotnet/jit-contrib.

File: src/jit/optcse.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/jit/optcse.cpp]: Still not clear to me if the above is a legality check followed by a profitability check, or if both are legality checks. Can you clarify?

Comment (briansull) [reply src/jit/optcse.cpp]: Currently is both a profitability check and a legality check.
This method is used in the hoisting logic and when it returns true we can perform an illegal hoist of a struct l-value.
I don't believe that creating CSE of multi-slot structs is ever going generate better code,
CSE are profitable when the def cost is expensive relative to the use costs.
The def costs for non-enregisterable  multi-slot structs is never going to be that much more expensive than the use cost.  If the def is  a heavy complex expression, such as an array expression we will still perform a CSE on the address calculation, leaving us with a simple multi-slot load as the CSE def. Since it isn't enregisterable  the CSE use still have to load from memory (the local stack frame) so we are just moving a copy of the struct from global memory into a stack copy and not saving any memory operations.

**Comment by briansull** [comment]
@dotnet-bot Test Windows_NT arm Cross Checked Innerloop Build and Test

**Comment by briansull** [comment]
On the desktop there are no reported Asm Diffs in the frameworks.

File: src/jit/optcse.cpp (hunk 1 above)
Comment (CarolEidt) [reply src/jit/optcse.cpp]: @briansull - in our discussions of this, it seemed like the fundamental issues are:
1) The struct handling in the JIT relies on the fact that a struct r-value will always be used in a context where its size and/or handle are carried by an l-value (e.g. a block node on the lhs of an assign, or a block store).
2) An unused hoisted rvalue doesn't get removed. This leaves the problematic handle-less size-less struct indirection.

Before taking this fix, I'd like us to understand what diffs this creates, and file an issue to address the above problems.

As we improve our struct codegen, I'd expect us to see more cases where hoisting struct values would be beneficial, as we can't (and probably never will) promote all structs that might be created redundantly (within a loop or otherwise).

Comment (mikedn) [reply src/jit/optcse.cpp]: > An unused hoisted rvalue doesn't get removed. This leaves the problematic handle-less size-less struct indirection.

Not only that it doesn't get removed but it simply cannot be used at all. Because loop hoisting is expecting CSE to assign the hoisted tree to a variable but CSE can't do that because there's struct type information available in the r-value.

To make this work with the current loop hoisiting implementation we'd need to stop discarding struct information from r-values (converting `OBJ` to `TYP_STRUCT` `IND`). That's doable and IMO desirable, discarding struct information is just risky.

Comment (AndyAyersMS) [reply src/jit/optcse.cpp]: FWIW I have seen many examples where manual struct CSEs are profitable -- for example the before version of this did not cache `Start` and `End` as local structs. IL semantics for structs tends to lead to lots of struct copies and often many of those copies are identical.

https://github.com/dotnet/coreclr/blob/0d581c79f0b40bd43ad3eb62574bfe78a665fe80/src/System.Private.CoreLib/shared/System/Range.cs#L105-L122

Comment (mikedn) [reply src/jit/optcse.cpp]: Hmm, should perhaps this fix be restricted to loop hoisting?

Comment (briansull) [reply src/jit/optcse.cpp]: I  have coded up an alternative fix that blocks the hoisting of TYP_STRUCT r-values.

**Comment by briansull** [comment]
I have coded up an alternative fix that blocks the hoisting of TYP_STRUCT r-values.
PTAL @CarolEidt @AndyAyersMS  @dotnet/jit-contrib.

Working on diffs

File: src/jit/optimizer.cpp [hunk 2]
```diff
@@ -6885,26 +6885,35 @@ bool Compiler::optHoistLoopExprsForTree(GenTree*          tree,
...
 
-        // If it's a call, it must be a helper call, and be pure.
-        // Further, if it may run a cctor, it must be labeled as "Hoistable"
-        // (meaning it won't run a cctor because the class is not precise-init).
-        if (treeIsHoistable && tree->OperGet() == GT_CALL)
+        if (treeIsHoistable)
         {
-            GenTreeCall* call = tree->AsCall();
-            if (call->gtCallType != CT_HELPER)
+            // We cannot hoist an r-value of TYP_STRUCT, as it is illegal to do so
```
Comment (CarolEidt) [inline src/jit/optimizer.cpp]: I would reword this comment as "We cannot hoist an r-value of TYP_STRUCT, as they currently do not carry full descriptors of the struct type" or something to that effect.

Comment (briansull) [reply src/jit/optimizer.cpp]: Done

**Comment by briansull** [comment]
Rebased and squashed commits

No Asm diffs
